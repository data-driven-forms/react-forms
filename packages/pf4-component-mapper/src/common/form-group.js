import React from 'react';
import { FormGroup as Pf4FormGroup, TextContent, Text } from '@patternfly/react-core';
import PropTypes from 'prop-types';

import showError from './show-error';

const FormGroup = ({ label, isRequired, helperText, meta, description, hideLabel, children, id }) => (
  <Pf4FormGroup
    isRequired={isRequired}
    label={!hideLabel && label}
    fieldId={id}
    helperText={helperText}
    helperTextInvalid={meta.error}
    {...showError(meta)}
  >
    {description && (
      <TextContent>
        <Text component="small">{description}</Text>
      </TextContent>
    )}
    {children}
  </Pf4FormGroup>
);

FormGroup.propTypes = {
  label: PropTypes.node,
  isRequired: PropTypes.bool,
  helperText: PropTypes.node,
  meta: PropTypes.object.isRequired,
  description: PropTypes.node,
  hideLabel: PropTypes.bool,
  id: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element)]).isRequired
};

FormGroup.defaultProps = {
  isRequired: false,
  description: undefined
};

export default FormGroup;
