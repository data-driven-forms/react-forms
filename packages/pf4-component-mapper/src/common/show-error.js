const showError = ({ error, touched }) => ({ validated: touched && error ? 'error' : 'default' });

export default showError;
