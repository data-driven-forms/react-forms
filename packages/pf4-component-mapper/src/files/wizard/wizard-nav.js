import React, { useState, useEffect } from 'react';

import { WizardNavItem, WizardNav } from '@patternfly/react-core';

import isEqual from 'lodash/isEqual';
import get from 'lodash/get';
import PropTypes from 'prop-types';

const memoValues = (initialValue) => {
  let valueCache = initialValue;

  return (value) => {
    if (!isEqual(value, valueCache)) {
      valueCache = value;
      return true;
    }

    return false;
  };
};

const WizardNavigationInternal = React.memo(
  ({ navSchema, activeStepIndex, maxStepIndex, jumpToStep, valid, validating }) =>
    navSchema
      .filter((field) => field.primary)
      .map((step) => {
        const substeps = step.substepOf && navSchema.filter((field) => field.substepOf === step.substepOf);

        const isValid = valid && !validating;

        return (
          <WizardNavItem
            key={step.substepOf || step.title}
            content={step.substepOf || step.title}
            isCurrent={substeps ? activeStepIndex >= step.index && activeStepIndex < step.index + substeps.length : activeStepIndex === step.index}
            isDisabled={isValid ? maxStepIndex < step.index : step.index > activeStepIndex}
            onNavItemClick={(ind) => jumpToStep(ind, isValid)}
            step={step.index}
          >
            {substeps && (
              <WizardNav returnList>
                {substeps.map((substep) => (
                  <WizardNavItem
                    key={substep.title}
                    content={substep.title}
                    isCurrent={activeStepIndex === substep.index}
                    isDisabled={isValid ? maxStepIndex < substep.index : substep.index > activeStepIndex}
                    onNavItemClick={(ind) => jumpToStep(ind, isValid)}
                    step={substep.index}
                  />
                ))}
              </WizardNav>
            )}
          </WizardNavItem>
        );
      }),
  isEqual
);

WizardNavigationInternal.propTypes = {
  activeStepIndex: PropTypes.number.isRequired,
  maxStepIndex: PropTypes.number.isRequired,
  jumpToStep: PropTypes.func.isRequired,
  navSchema: PropTypes.array.isRequired,
  valid: PropTypes.bool.isRequired,
  validating: PropTypes.bool.isRequired
};

const WizardNavigation = ({ setPrevSteps, crossroads, values, ...props }) => {
  const [memoValue] = useState(() =>
    memoValues(
      crossroads
        ? crossroads.reduce(
            (acc, curr) => ({
              ...acc,
              [curr]: get(values, curr)
            }),
            {}
          )
        : {}
    )
  );

  useEffect(() => {
    if (crossroads) {
      const modifiedRoad = crossroads.reduce(
        (acc, curr) => ({
          ...acc,
          [curr]: get(values, curr)
        }),
        {}
      );

      if (memoValue(modifiedRoad)) {
        setPrevSteps();
      }
    }
  });

  return <WizardNavigationInternal {...props} />;
};

WizardNavigation.propTypes = {
  setPrevSteps: PropTypes.func.isRequired,
  values: PropTypes.object.isRequired,
  crossroads: PropTypes.arrayOf(PropTypes.string)
};

export default WizardNavigation;
