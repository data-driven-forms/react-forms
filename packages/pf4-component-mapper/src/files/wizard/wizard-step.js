import React, { Fragment, useEffect, useRef } from 'react';
import { Title, WizardBody } from '@patternfly/react-core';
import PropTypes from 'prop-types';
import WizardStepButtons from './step-buttons';

export const RenderTitle = ({ title, customTitle }) =>
  customTitle ? (
    customTitle
  ) : (
    <Title headingLevel="h1" size="xl">
      {title}
    </Title>
  );

RenderTitle.propTypes = {
  title: PropTypes.node,
  customTitle: PropTypes.node
};

const WizardStep = ({ name, title, description, fields, formOptions, showTitles, showTitle, customTitle, hasNoBodyPadding, ...rest }) => {
  const formRef = useRef();

  useEffect(() => {
    // HACK: I can not pass ref to WizardBody because it is not
    // wrapped by forwardRef. However, the step body (the one that overflows)
    // is the grand parent of the form element.
    const stepBody = formRef.current && formRef.current.parentNode.parentNode;
    stepBody.scrollTo({ top: 0, left: 0, behavior: 'smooth' });
  }, [name]);

  return (
    <Fragment>
      <WizardBody hasNoBodyPadding={hasNoBodyPadding}>
        <div ref={formRef} className="pf-c-form">
          {((showTitles && showTitle !== false) || showTitle) && <RenderTitle title={title} customTitle={customTitle} />}
          {fields.map((item) => formOptions.renderForm([item], formOptions))}
        </div>
      </WizardBody>
      <WizardStepButtons formOptions={formOptions} {...rest} />
    </Fragment>
  );
};

WizardStep.propTypes = {
  title: PropTypes.node,
  description: PropTypes.node,
  fields: PropTypes.array.isRequired,
  formOptions: PropTypes.shape({
    renderForm: PropTypes.func.isRequired
  }).isRequired,
  showTitles: PropTypes.bool,
  showTitle: PropTypes.bool,
  customTitle: PropTypes.node,
  name: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  hasNoBodyPadding: PropTypes.bool
};

export default WizardStep;
