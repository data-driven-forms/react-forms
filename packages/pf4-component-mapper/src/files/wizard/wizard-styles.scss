.no-shadow {
  box-shadow: none;
}

.pull-right {
  justify-content: flex-end;
}

.ddorg__pf4-component-mapper__wizard {
  height: calc(100vh - 48px);
  max-height: 762px;
}
