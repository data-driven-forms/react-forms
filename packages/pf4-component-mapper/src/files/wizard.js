import React, { useReducer, useEffect } from 'react';
import { createPortal } from 'react-dom';
import PropTypes from 'prop-types';
import { FormSpy } from '@data-driven-forms/react-form-renderer';
import Wizard, { wizardProps } from '@data-driven-forms/common/src/wizard/wizard';

import { Bullseye, Backdrop, WizardNav, WizardHeader } from '@patternfly/react-core';

import WizardStep from './wizard/wizard-step';
import './wizard/wizard-styles.scss';

import WizardNavigation from './wizard/wizard-nav';
import reducer from './wizard/reducer';

const Modal = ({ children, container, inModal }) =>
  inModal
    ? createPortal(
        <Backdrop>
          <Bullseye>
            <div className="pf-c-modal-box pf-m-lg">{children}</div>
          </Bullseye>
        </Backdrop>,
        container
      )
    : children;

const WizardInternal = ({
  inModal,
  crossroads,
  title,
  description,
  buttonLabels,
  buttonsClassName,
  showTitles,
  formOptions,
  currentStep,
  handlePrev,
  onKeyDown,
  jumpToStep,
  setPrevSteps,
  handleNext,
  navSchema,
  activeStepIndex,
  maxStepIndex,
  isDynamic,
  container,
  hideClose,
  titleId,
  descriptionId,
  closeButtonAriaLabel,
  hasNoBodyPadding,
  prevSteps,
  navAriaLabel,
  ...rest
}) => {
  const [state, dispatch] = useReducer(reducer, { loading: true, container });

  useEffect(() => {
    if (inModal) {
      dispatch({ type: 'setContainer' });
    } else {
      dispatch({ type: 'finishLoading' });
    }
  }, [inModal]);

  useEffect(() => {
    if (state.container) {
      document.body.appendChild(state.container);
      dispatch({ type: 'finishLoading' });
    }

    return () => {
      if (inModal && state.container) {
        document.body.removeChild(state.container);
      }
    };
  }, [state.container, inModal]);

  if (state.loading) {
    return null;
  }

  return (
    <Modal inModal={inModal} container={state.container}>
      <div
        className={`pf-c-wizard ${inModal ? '' : 'no-shadow'}`}
        role="dialog"
        aria-modal={inModal ? 'true' : undefined}
        onKeyDown={onKeyDown}
        {...rest}
      >
        {title && (
          <WizardHeader
            title={title}
            description={description}
            onClose={formOptions.onCancel}
            hideClose={hideClose}
            titleId={titleId}
            descriptionId={descriptionId}
            closeButtonAriaLabel={closeButtonAriaLabel}
          />
        )}
        <div className="pf-c-wizard__outer-wrap">
          <div className="pf-c-wizard__inner-wrap">
            <WizardNav aria-label={navAriaLabel}>
              <FormSpy subscription={{ values: true, valid: true, validating: true }}>
                {({ values, valid, validating }) => (
                  <WizardNavigation
                    navSchema={navSchema}
                    activeStepIndex={activeStepIndex}
                    valid={valid}
                    maxStepIndex={maxStepIndex}
                    jumpToStep={jumpToStep}
                    crossroads={crossroads}
                    isDynamic={isDynamic}
                    values={values}
                    setPrevSteps={setPrevSteps}
                    validating={validating}
                  />
                )}
              </FormSpy>
            </WizardNav>
            <WizardStep
              {...currentStep}
              formOptions={formOptions}
              buttonLabels={buttonLabels}
              buttonsClassName={buttonsClassName}
              showTitles={showTitles}
              handleNext={(nextStep) => handleNext(nextStep)}
              handlePrev={handlePrev}
              disableBack={activeStepIndex === 0}
              hasNoBodyPadding={hasNoBodyPadding}
            />
          </div>
        </div>
      </div>
    </Modal>
  );
};

WizardInternal.propTypes = {
  buttonLabels: PropTypes.shape({
    submit: PropTypes.node.isRequired,
    cancel: PropTypes.node.isRequired,
    back: PropTypes.node.isRequired,
    next: PropTypes.node.isRequired
  }).isRequired,
  buttonsClassName: PropTypes.string,
  title: PropTypes.any,
  description: PropTypes.any,
  isCompactNav: PropTypes.bool,
  inModal: PropTypes.bool,
  setFullWidth: PropTypes.bool,
  setFullHeight: PropTypes.bool,
  isDynamic: PropTypes.bool,
  showTitles: PropTypes.bool,
  crossroads: PropTypes.arrayOf(PropTypes.string),
  hideClose: PropTypes.bool,
  titleId: PropTypes.string,
  descriptionId: PropTypes.string,
  closeButtonAriaLabel: PropTypes.string,
  hasNoBodyPadding: PropTypes.bool,
  navAriaLabel: PropTypes.string,
  ...wizardProps
};

const defaultLabels = {
  submit: 'Submit',
  cancel: 'Cancel',
  back: 'Back',
  next: 'Next'
};

const WizardFunction = ({ buttonLabels, ...props }) => (
  <Wizard Wizard={WizardInternal} {...props} buttonLabels={{ ...defaultLabels, ...buttonLabels }} />
);

WizardFunction.propTypes = {
  buttonLabels: PropTypes.shape({
    submit: PropTypes.node,
    cancel: PropTypes.node,
    back: PropTypes.node
  })
};

WizardFunction.defaultProps = {
  buttonLabels: {}
};

export default WizardFunction;
