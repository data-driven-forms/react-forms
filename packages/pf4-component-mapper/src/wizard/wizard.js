import React, { useReducer, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import { FormSpy, WizardContext } from '@data-driven-forms/react-form-renderer';
import Wizard from '@data-driven-forms/common/wizard/wizard';

import { WizardNav, WizardHeader, Modal as PF4Modal } from '@patternfly/react-core';

import WizardStep from './wizard-components/wizard-step';
import './wizard-components/wizard-styles.css';

import WizardNavigation from './wizard-components/wizard-nav';
import reducer from './wizard-components/reducer';
import WizardToggle from './wizard-components/wizard-toggle';

const Modal = ({ children, container, inModal, ...rest }) =>
  inModal ? (
    <PF4Modal variant="large" isOpen showClose={false} hasNoBodyWrapper appendTo={container} {...rest}>
      {children}
    </PF4Modal>
  ) : (
    children
  );

Modal.propTypes = {
  children: PropTypes.node,
  container: PropTypes.instanceOf(Element),
  inModal: PropTypes.bool,
};

const WizardInternal = ({
  inModal,
  title,
  description,
  buttonLabels,
  buttonsClassName,
  showTitles,
  container,
  hideClose,
  titleId,
  descriptionId,
  closeButtonAriaLabel,
  hasNoBodyPadding,
  navAriaLabel,
  StepTemplate,
  className,
  conditionalSubmitFlag,
  ModalProps = {},
  ...rest
}) => {
  const {
    crossroads,
    formOptions,
    currentStep,
    handlePrev,
    onKeyDown,
    jumpToStep,
    setPrevSteps,
    handleNext,
    navSchema,
    activeStepIndex,
    maxStepIndex,
    isDynamic,
  } = useContext(WizardContext);
  const [state, dispatch] = useReducer(reducer, { loading: true, container, openNav: false });

  useEffect(() => {
    if (inModal) {
      dispatch({ type: 'setContainer' });
    } else {
      dispatch({ type: 'finishLoading' });
    }
  }, [inModal]);

  useEffect(() => {
    if (ModalProps.appendTo) {
      // override with default append to
      dispatch({ type: 'finishLoading' });
    } else if (state.container) {
      document.body.appendChild(state.container);
      dispatch({ type: 'finishLoading' });
    }

    return () => {
      if (inModal && state.container) {
        document.body.removeChild(state.container);
      }
    };
  }, [state.container, inModal]);

  if (state.loading) {
    return null;
  }

  return (
    <Modal inModal={inModal} container={state.container} aria-labelledby={rest.name} {...ModalProps}>
      <div
        tabIndex={inModal ? 0 : null}
        className={`pf-c-wizard ${inModal ? '' : 'no-shadow'} ddorg__pf4-component-mapper__wizard ${className ? className : ''}`}
        role="dialog"
        aria-modal={inModal ? 'true' : undefined}
        onKeyDown={(e) => {
          onKeyDown(e);
          if (e.key === 'Escape' && inModal) {
            formOptions.onCancel();
          }
        }}
        {...rest}
      >
        {title && (
          <WizardHeader
            title={title}
            description={description}
            onClose={formOptions.onCancel}
            hideClose={hideClose}
            titleId={titleId}
            descriptionId={descriptionId}
            closeButtonAriaLabel={closeButtonAriaLabel}
          />
        )}
        <WizardToggle activeStepIndex={activeStepIndex} currentStep={currentStep} navSchema={navSchema} isOpen={state.openNav} dispatch={dispatch} />
        <div className="pf-c-wizard__outer-wrap">
          <div className="pf-c-wizard__inner-wrap">
            <WizardNav aria-label={navAriaLabel} isOpen={state.openNav}>
              <FormSpy subscription={{ values: true, valid: true, validating: true }}>
                {({ values, valid, validating }) => (
                  <WizardNavigation
                    navSchema={navSchema}
                    activeStepIndex={activeStepIndex}
                    valid={valid}
                    maxStepIndex={maxStepIndex}
                    jumpToStep={(...args) => {
                      state.openNav && dispatch({ type: 'closeNav' });
                      return jumpToStep(...args);
                    }}
                    crossroads={crossroads}
                    isDynamic={isDynamic}
                    values={values}
                    setPrevSteps={setPrevSteps}
                    validating={validating}
                  />
                )}
              </FormSpy>
            </WizardNav>
            <WizardStep
              conditionalSubmitFlag={conditionalSubmitFlag}
              buttonLabels={buttonLabels}
              buttonsClassName={buttonsClassName}
              showTitles={showTitles}
              hasNoBodyPadding={hasNoBodyPadding}
              StepTemplate={StepTemplate}
              {...currentStep}
              formOptions={formOptions}
              handleNext={(nextStep) => handleNext(nextStep)}
              handlePrev={handlePrev}
              disableBack={activeStepIndex === 0}
            />
          </div>
        </div>
      </div>
    </Modal>
  );
};

WizardInternal.propTypes = {
  buttonLabels: PropTypes.shape({
    submit: PropTypes.node.isRequired,
    cancel: PropTypes.node.isRequired,
    back: PropTypes.node.isRequired,
    next: PropTypes.node.isRequired,
  }).isRequired,
  buttonsClassName: PropTypes.string,
  title: PropTypes.any,
  description: PropTypes.any,
  inModal: PropTypes.bool,
  showTitles: PropTypes.bool,
  hideClose: PropTypes.bool,
  titleId: PropTypes.string,
  descriptionId: PropTypes.string,
  closeButtonAriaLabel: PropTypes.string,
  hasNoBodyPadding: PropTypes.bool,
  navAriaLabel: PropTypes.string,
  container: PropTypes.instanceOf(Element),
  StepTemplate: PropTypes.elementType,
  className: PropTypes.string,
  conditionalSubmitFlag: PropTypes.string,
  ModalProps: PropTypes.object,
};

const defaultLabels = {
  submit: 'Submit',
  cancel: 'Cancel',
  back: 'Back',
  next: 'Next',
};

const WizardFunction = ({ buttonLabels, ...props }) => (
  <Wizard Wizard={WizardInternal} {...props} buttonLabels={{ ...defaultLabels, ...buttonLabels }} />
);

WizardFunction.propTypes = {
  buttonLabels: PropTypes.shape({
    submit: PropTypes.node,
    cancel: PropTypes.node,
    back: PropTypes.node,
  }),
};

WizardFunction.defaultProps = {
  buttonLabels: {},
};

export default WizardFunction;
