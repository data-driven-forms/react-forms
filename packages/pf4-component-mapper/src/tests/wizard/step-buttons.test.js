import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import handleEnter from '@data-driven-forms/common/wizard/enter-handler';

import RenderWithProvider from '../../../../../__mocks__/with-provider';

import WizardStepButtons from '../../wizard/wizard-components/step-buttons';
import componentTypes from '@data-driven-forms/react-form-renderer/component-types';
import { CONDITIONAL_SUBMIT_FLAG } from '@data-driven-forms/common/wizard';
import validatorTypes from '@data-driven-forms/react-form-renderer/validator-types';
import { FormRenderer } from '@data-driven-forms/react-form-renderer';
import FormTemplate from '../../form-template';
import componentMapper from '../../component-mapper';

describe('<WizardSTepButtons', () => {
  let initialProps;
  beforeEach(() => {
    initialProps = {
      conditionalSubmitFlag: 'submit',
      buttonLabels: {
        cancel: 'Cancel',
        next: 'Next',
        submit: 'Submit',
        back: 'Back',
      },
      formOptions: {
        onCancel: jest.fn(),
        handleSubmit: jest.fn(),
        submit: jest.fn(),
        valid: true,
        getState: () => ({
          validating: false,
        }),
      },
      handlePrev: jest.fn(),
      handleNext: jest.fn(),
    };
  });

  it('should render correctly', () => {
    render(
      <RenderWithProvider>
        <WizardStepButtons {...initialProps} />
      </RenderWithProvider>
    );

    expect(screen.getByText('Cancel')).toBeInTheDocument();
    expect(screen.getByText('Submit')).toBeInTheDocument();
    expect(screen.getByText('Back')).toBeInTheDocument();
  });

  it('should add custom className to toolbar', () => {
    render(
      <RenderWithProvider>
        <WizardStepButtons {...initialProps} buttonsClassName="foo-class" />
      </RenderWithProvider>
    );

    expect(screen.getByText('Cancel').closest('.foo-class')).toBeInTheDocument();
  });

  it('should call next with correct arguments when next step is string', async () => {
    const handleNext = jest.fn();

    render(
      <RenderWithProvider>
        <WizardStepButtons {...initialProps} handleNext={handleNext} nextStep="next-step" />
      </RenderWithProvider>
    );

    await userEvent.click(screen.getByText('Next'));

    expect(handleNext).toHaveBeenCalledWith('next-step');
  });

  it('should call next with correct arguments when next step is condition', async () => {
    const handleNext = jest.fn();

    render(
      <RenderWithProvider>
        <WizardStepButtons
          {...initialProps}
          handleNext={handleNext}
          formOptions={{
            ...initialProps.formOptions,
            getState: () => ({ values: { foo: 'foo' } }),
          }}
          nextStep={{
            when: 'foo',
            stepMapper: {
              foo: 'bar',
              qux: 'quaxx',
            },
          }}
        />
      </RenderWithProvider>
    );

    await userEvent.click(screen.getByText('Next'));

    expect(handleNext).toHaveBeenCalledWith('bar');
  });

  it('should call submit functions if no next step is defined', async () => {
    const handleSubmit = jest.fn();
    render(
      <RenderWithProvider>
        <WizardStepButtons {...initialProps} formOptions={{ ...initialProps.formOptions, handleSubmit }} nextStep={undefined} />
      </RenderWithProvider>
    );

    await userEvent.click(screen.getByText('Submit'));

    expect(handleSubmit).toHaveBeenCalled();
  });

  it('should call cancel function', async () => {
    const VALUES = { aws: 'yes', password: '123456643' };
    const onCancel = jest.fn();
    render(
      <RenderWithProvider>
        <WizardStepButtons {...initialProps} formOptions={{ ...initialProps.formOptions, onCancel, getState: () => ({ values: VALUES }) }} />
      </RenderWithProvider>
    );

    await userEvent.click(screen.getByText('Cancel'));

    expect(onCancel).toHaveBeenCalled();
  });

  it('should call prev function', async () => {
    const handlePrev = jest.fn();
    render(
      <RenderWithProvider>
        <WizardStepButtons {...initialProps} handlePrev={handlePrev} disableBack={false} />
      </RenderWithProvider>
    );

    await userEvent.click(screen.getByText('Back'));

    expect(handlePrev).toHaveBeenCalled();
  });

  describe('.handleEnter', () => {
    let event;
    let formOptions;
    let activeStep;
    let findCurrentStep;
    let handleNext;
    let handleSubmit;

    let nextStep;
    let getRegisteredFields;

    beforeEach(() => {
      nextStep = 'cosi';
      getRegisteredFields = jest.fn();

      event = {
        target: {
          type: 'notbutton',
        },
        key: 'Enter',
        preventDefault: jest.fn(),
      };
      formOptions = {
        valid: true,
        getRegisteredFields,
        getState: () => ({ validating: false }),
      };
      activeStep = 'active-step';
      findCurrentStep = jest.fn().mockImplementation(() => ({ nextStep }));
      handleNext = jest.fn();
      handleSubmit = jest.fn();
    });

    it('should call next', () => {
      handleEnter(event, formOptions, activeStep, findCurrentStep, handleNext, handleSubmit);

      expect(event.preventDefault).toHaveBeenCalled();
      expect(handleNext).toHaveBeenCalledWith(nextStep, getRegisteredFields);
      expect(handleSubmit).not.toHaveBeenCalled();
    });

    it('should call submit', () => {
      findCurrentStep = jest.fn().mockImplementation(() => ({}));

      handleEnter(event, formOptions, activeStep, findCurrentStep, handleNext, handleSubmit);

      expect(event.preventDefault).toHaveBeenCalled();
      expect(handleNext).not.toHaveBeenCalled();
      expect(handleSubmit).toHaveBeenCalled();
    });

    it('should be prevented if step has custom buttons', () => {
      findCurrentStep = jest.fn().mockImplementation(() => ({ buttons: true }));

      handleEnter(event, formOptions, activeStep, findCurrentStep, handleNext, handleSubmit);

      expect(event.preventDefault).toHaveBeenCalled();
      expect(handleNext).not.toHaveBeenCalled();
      expect(handleSubmit).not.toHaveBeenCalled();
    });

    it('should be prevented if form is not valid', () => {
      formOptions = {
        valid: false,
        getRegisteredFields,
        getState: () => ({ validating: false }),
      };

      handleEnter(event, formOptions, activeStep, findCurrentStep, handleNext, handleSubmit);

      expect(event.preventDefault).toHaveBeenCalled();
      expect(handleNext).not.toHaveBeenCalled();
      expect(handleSubmit).not.toHaveBeenCalled();
    });

    it('should be prevented if form is validating', () => {
      formOptions = {
        valid: true,
        getRegisteredFields,
        getState: () => ({ validating: true }),
      };

      handleEnter(event, formOptions, activeStep, findCurrentStep, handleNext, handleSubmit);

      expect(event.preventDefault).toHaveBeenCalled();
      expect(handleNext).not.toHaveBeenCalled();
      expect(handleSubmit).not.toHaveBeenCalled();
    });

    it('nothing happen when key is not enter', () => {
      event = {
        ...event,
        key: 'Spacebar',
      };

      handleEnter(event, formOptions, activeStep, findCurrentStep, handleNext, handleSubmit);

      expect(event.preventDefault).not.toHaveBeenCalled();
      expect(handleNext).not.toHaveBeenCalled();
      expect(handleSubmit).not.toHaveBeenCalled();
    });

    it('nothing happen when ctrl is pressed', () => {
      event = {
        ...event,
        ctrlKey: true,
      };

      handleEnter(event, formOptions, activeStep, findCurrentStep, handleNext, handleSubmit);

      expect(event.preventDefault).not.toHaveBeenCalled();
      expect(handleNext).not.toHaveBeenCalled();
      expect(handleSubmit).not.toHaveBeenCalled();
    });

    it('nothing happen when shift is pressed', () => {
      event = {
        ...event,
        shiftKey: true,
      };

      handleEnter(event, formOptions, activeStep, findCurrentStep, handleNext, handleSubmit);

      expect(event.preventDefault).not.toHaveBeenCalled();
      expect(handleNext).not.toHaveBeenCalled();
      expect(handleSubmit).not.toHaveBeenCalled();
    });

    it('nothing happen when target is button', () => {
      event = {
        ...event,
        target: {
          ...event.target,
          type: 'button',
        },
      };

      handleEnter(event, formOptions, activeStep, findCurrentStep, handleNext, handleSubmit);

      expect(event.preventDefault).not.toHaveBeenCalled();
      expect(handleNext).not.toHaveBeenCalled();
      expect(handleSubmit).not.toHaveBeenCalled();
    });
  });

  it('conditional submit step', async () => {
    const submit = jest.fn();
    const schema = {
      fields: [
        {
          component: componentTypes.WIZARD,
          name: 'wizard',
          fields: [
            {
              name: 'first-step',
              nextStep: {
                when: 'name',
                stepMapper: {
                  aws: 'summary',
                  submit: CONDITIONAL_SUBMIT_FLAG,
                },
              },
              fields: [
                {
                  component: componentTypes.TEXT_FIELD,
                  name: 'name',
                  validate: [{ type: validatorTypes.REQUIRED }],
                  label: 'Name',
                },
              ],
            },
          ],
        },
      ],
    };

    render(
      <FormRenderer
        componentMapper={componentMapper}
        FormTemplate={(props) => <FormTemplate {...props} showFormControls={false} />}
        onSubmit={submit}
        schema={schema}
      />
    );

    await userEvent.type(screen.getByLabelText('Name'), 'bla');

    expect(screen.getByText('Next')).toBeInTheDocument();
    expect(() => screen.getByText('Submit')).toThrow();

    await userEvent.clear(screen.getByLabelText('Name'));
    await userEvent.type(screen.getByLabelText('Name'), 'submit');

    expect(() => screen.getByText('Next')).toThrow();
    expect(screen.getByText('Submit')).toBeInTheDocument();

    await userEvent.click(screen.getByText('Submit'));

    expect(submit).toHaveBeenCalledWith({ name: 'submit' }, expect.any(Object), expect.any(Object));
  });
});
