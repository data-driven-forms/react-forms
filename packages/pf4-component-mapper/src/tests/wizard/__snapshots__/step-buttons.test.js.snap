// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WizardSTepButtons should add custom className to toolbar 1`] = `
<RenderWithProvider>
  <ReactFinalForm
    onSubmit={[Function]}
  >
    <WizardStepButtons
      buttonLabels={
        Object {
          "back": "Back",
          "cancel": "Cancel",
          "next": "Next",
          "submit": "Submit",
        }
      }
      buttonsClassName="foo-class"
      formOptions={
        Object {
          "getState": [Function],
          "handleSubmit": [MockFunction],
          "onCancel": [MockFunction],
          "submit": [MockFunction],
          "valid": true,
        }
      }
      handleNext={[MockFunction]}
      handlePrev={[MockFunction]}
    >
      <footer
        className="pf-c-wizard__footer foo-class"
      >
        <FormSpy>
          <NextButton
            getState={[Function]}
            handleNext={[MockFunction]}
            handleSubmit={[MockFunction]}
            nextLabel="Next"
            onCancel={[MockFunction]}
            submit={[MockFunction]}
            submitLabel="Submit"
            valid={true}
          >
            <Component
              isDisabled={false}
              onClick={[Function]}
              type="button"
              variant="primary"
            >
              <button
                aria-disabled={null}
                aria-label={null}
                className="pf-c-button pf-m-primary"
                data-ouia-component-id={null}
                data-ouia-component-type="PF4/Button"
                data-ouia-safe={true}
                disabled={false}
                onClick={[Function]}
                tabIndex={null}
                type="button"
              >
                Submit
              </button>
            </Component>
          </NextButton>
          <Component
            onClick={[MockFunction]}
            type="button"
            variant="secondary"
          >
            <button
              aria-disabled={null}
              aria-label={null}
              className="pf-c-button pf-m-secondary"
              data-ouia-component-id={null}
              data-ouia-component-type="PF4/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[MockFunction]}
              tabIndex={null}
              type="button"
            >
              Back
            </button>
          </Component>
          <Component
            onClick={[MockFunction]}
            type="button"
            variant="link"
          >
            <button
              aria-disabled={null}
              aria-label={null}
              className="pf-c-button pf-m-link"
              data-ouia-component-id={null}
              data-ouia-component-type="PF4/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[MockFunction]}
              tabIndex={null}
              type="button"
            >
              Cancel
            </button>
          </Component>
        </FormSpy>
      </footer>
    </WizardStepButtons>
  </ReactFinalForm>
</RenderWithProvider>
`;

exports[`<WizardSTepButtons should render correctly 1`] = `
<RenderWithProvider>
  <ReactFinalForm
    onSubmit={[Function]}
  >
    <WizardStepButtons
      buttonLabels={
        Object {
          "back": "Back",
          "cancel": "Cancel",
          "next": "Next",
          "submit": "Submit",
        }
      }
      formOptions={
        Object {
          "getState": [Function],
          "handleSubmit": [MockFunction],
          "onCancel": [MockFunction],
          "submit": [MockFunction],
          "valid": true,
        }
      }
      handleNext={[MockFunction]}
      handlePrev={[MockFunction]}
    >
      <footer
        className="pf-c-wizard__footer "
      >
        <FormSpy>
          <NextButton
            getState={[Function]}
            handleNext={[MockFunction]}
            handleSubmit={[MockFunction]}
            nextLabel="Next"
            onCancel={[MockFunction]}
            submit={[MockFunction]}
            submitLabel="Submit"
            valid={true}
          >
            <Component
              isDisabled={false}
              onClick={[Function]}
              type="button"
              variant="primary"
            >
              <button
                aria-disabled={null}
                aria-label={null}
                className="pf-c-button pf-m-primary"
                data-ouia-component-id={null}
                data-ouia-component-type="PF4/Button"
                data-ouia-safe={true}
                disabled={false}
                onClick={[Function]}
                tabIndex={null}
                type="button"
              >
                Submit
              </button>
            </Component>
          </NextButton>
          <Component
            onClick={[MockFunction]}
            type="button"
            variant="secondary"
          >
            <button
              aria-disabled={null}
              aria-label={null}
              className="pf-c-button pf-m-secondary"
              data-ouia-component-id={null}
              data-ouia-component-type="PF4/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[MockFunction]}
              tabIndex={null}
              type="button"
            >
              Back
            </button>
          </Component>
          <Component
            onClick={[MockFunction]}
            type="button"
            variant="link"
          >
            <button
              aria-disabled={null}
              aria-label={null}
              className="pf-c-button pf-m-link"
              data-ouia-component-id={null}
              data-ouia-component-type="PF4/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[MockFunction]}
              tabIndex={null}
              type="button"
            >
              Cancel
            </button>
          </Component>
        </FormSpy>
      </footer>
    </WizardStepButtons>
  </ReactFinalForm>
</RenderWithProvider>
`;
